File: Octave/metodo_QR.m
##Metodo QR para el calculo de autovalores
##Entradas: Matriz y tolerancia minima
##Salidas: Valores propios de la matriz
function A = metodo_QR(A, tol)
    function [Q, R] = factorizacion_QR(A1)
       function result =  producto_punto(x, y)
            result = 0;
            for k = 1:length(x)
                result += x(k) * y(k);
            endfor
       endfunction
       Q = zeros(length(A1));
       for i = 1:length(A1)
            Ak = A1(:, i);
            if i == 1
                U = Ak;
            else
                sum = 0;
                for j = 1:i-1
                    sum += producto_punto(Ak, Q(:,j)) * Q(:,j);
                endfor
                U = Ak - sum;
            end
            Q(:,i) = U / norm(U);
        endfor
        R = transpose(Q) * A1;
        Q = Q;
    endfunction
    error = tol;
    while error >= tol
        [Q, R] = factorizacion_QR(A);
        A = R * Q;
        error = 0;
        for i = 2:length(A)-1
            for j = 1:i-1
                error += abs(A(i, j));
            endfor
        endfor
    endwhile
    A = diag(A);
endfunction

File: Python/metodo_QR.py
import numpy as np


def producto_punto(x, y):
    result = 0
    for i in range(len(x)):
        result += x[i] * y[i]
    return result


def factorizacion_QR(A):
    A = np.array(A)
    Q = []
    for i in range(len(A)):
        Ak = A[:, i]
        if i == 0:
            U = Ak
        else:
            sum = 0
            for j in range(i):
                sum += producto_punto(Ak, Q[j]) * np.array(Q[j])
            U = A[:, i] - sum
        P = U / np.linalg.norm(U)
        Q.append(P.tolist())
    R = np.dot(Q, A)
    Q = np.transpose(Q)
    return Q, R


"""
Metodo QR para el calculo de autovalores
Entradas: Matriz y tolerancia minima
Salidas: Valores propios de la matriz
"""
def metodo_QR(A, tol):
    error = tol
    while error >= tol:
        QR = factorizacion_QR(A)
        A = np.dot(QR[1], QR[0])
        error = 0
        for i in range(1, len(A)):
            for j in range(i):
                error += abs(A[i][j])
    return np.diagonal(A)


print(metodo_QR([[3, 1, 0], [1, 3, 1], [0, 1, 3]], 0.00001))
